File: java/src/main/java/ezocalc/structure/Operation.java
package ezocalc.structure;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

public class Operation implements Bloc {

	List<Bloc> blocs = new ArrayList<>();

	public void append(Bloc b) {
		blocs.add(b);
	}

	public BigDecimal getValue() throws SyntaxError {
		// Returns value of the bloc (direct value or calculation)

		int pos = findOpe();

		while (pos != -1) {
			Operand ope = (Operand) blocs.get(pos);

			BigDecimal val = ope.calculate(blocs.get(pos - 1), blocs.get(pos + 1));
			blocs.set(pos - 1, new Value(val));
			blocs.remove(pos);
			blocs.remove(pos);

			pos = findOpe();
		}

		return blocs.get(0).getValue();
	}

	private Integer findOpe() {

		for (int i = 0; i < blocs.size(); i++) {
			Bloc b = blocs.get(i);
			if ((b instanceof Operand) && ((Operand) b).isFirst()) {
				return i;
			}
		}

		for (int i = 0; i < blocs.size(); i++) {
			Bloc b = blocs.get(i);
			if (b instanceof Operand && ((Operand) b).isSecond()) {
				return i;
			}
		}

		for (int i = 0; i < blocs.size(); i++) {
			Bloc b = blocs.get(i);
			if (b instanceof Operand) {
				return i;
			}
		}

		return -1;
	}

}

File: lambda/lambdas/ezocalc/operation.py
#!/usr/bin/env python3

from numbers import Number
import ezocalc.bloc
import ezocalc.operand
import ezocalc.value

class Operation(ezocalc.bloc.Bloc):

    _blocs = None

    def __init__(self):
        self._blocs = []

    def append(self, b:ezocalc.bloc.Bloc):
        self._blocs.append(b)

    def get_value(self) -> Number:
        """Returns value of the bloc (direct value or calculation)"""

        (pos, ope) = self._find_ope()
        while pos > 0:
            val = ope.calculate(self._blocs[pos-1], self._blocs[pos+1])
            self._blocs[pos-1] = ezocalc.value.Value(val)
            del self._blocs[pos:pos+2]

            (pos, ope) = self._find_ope()

        return self._blocs[0].get_value()


    def _find_ope(self):
        pos = 0
        for e in self._blocs:
            if type(e) is ezocalc.operand.Operand and e.is_first():
                return (pos, e)
            pos += 1

        pos = 0
        for e in self._blocs:
            if type(e) is ezocalc.operand.Operand and e.is_second():
                return (pos, e)
            pos += 1

        pos = 0
        for e in self._blocs:
            if type(e) is ezocalc.operand.Operand:
                return (pos, e)
            pos += 1

        return (-1, None)
